### PR TITLE
Splunk Observability Scaler - Initial PR

### PR DESCRIPTION
**Pull request by sschimper-splunk**
_Provide a description of what has been changed_

### Checklist

- [ ] When introducing a new scaler, I agree with the [scaling governance policy](https://github.com/kedacore/governance/blob/main/SCALERS.md)
- [ ] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [ ] Tests have been added
- [ ] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [ ] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [ ] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [ ] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

Fixes #

Relates to #

### CODE REVIEW

File: pkg/metricsservice/api/metrics.pb.go [hunk 1]
```diff
@@ -16,7 +16,7 @@
 // Code generated by protoc-gen-go. DO NOT EDIT.
 // versions:
 // 	protoc-gen-go v1.32.0
-// 	protoc        v4.23.2
+// 	protoc        v3.6.1
```
Comment (iwankgb) [inline pkg/metricsservice/api/metrics.pb.go:19]: Are you using outdated version of `protoc`? You shall not, undo changes to all the `.pb.go` files, please.

File: pkg/scalers/liiklus/LiiklusService.pb.go [hunk 2]
```diff
@@ -1,16 +1,16 @@
 // Code generated by protoc-gen-go. DO NOT EDIT.
 // versions:
 // 	protoc-gen-go v1.32.0
-// 	protoc        v4.23.2
```
Comment (iwankgb) [inline pkg/scalers/liiklus/LiiklusService.pb.go:4]: This file is unrelated to your changes, isn't it? These changes should not be here, I believe.

File: pkg/scalers/liiklus/LiiklusService_grpc.pb.go [hunk 3]
```diff
@@ -1,17 +1,17 @@
 // Code generated by protoc-gen-go-grpc. DO NOT EDIT.
 // versions:
 // - protoc-gen-go-grpc v1.3.0
-// - protoc             v4.23.2
```
Comment (iwankgb) [inline pkg/scalers/liiklus/LiiklusService_grpc.pb.go:4]: These changes are not related to your scaler too, I think.

File: pkg/scalers/splunk_o11y_scaler.go [hunk 4]
```diff
@@ -0,0 +1,264 @@
+package scalers
+
+import (
+	"context"
+	"fmt"
+	"math"
+
+	//"os"
```
Comment (iwankgb) [inline pkg/scalers/splunk_o11y_scaler.go:8]: Remove commented out imports, please :)

File: pkg/scalers/splunk_o11y_scaler.go [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-logr/logr"
+	// "github.com/signalfx/signalfx-go/signalflow/v2"
```
Comment (iwankgb) [inline pkg/scalers/splunk_o11y_scaler.go:14]: Ditto.

File: pkg/scalers/splunk_o11y_scaler.go [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+	// queryValue
+	if val, ok := config.TriggerMetadata["queryValue"]; ok {
+		queryValue, err := strconv.ParseFloat(val, 64)
+		if err != nil {
+			return nil, fmt.Errorf("queryValue parsing error %w", err)
+		}
+		meta.queryValue = queryValue
+	} else {
+		if config.AsMetricSource {
+			meta.queryValue = 0
```
Comment (iwankgb) [inline pkg/scalers/splunk_o11y_scaler.go:86]: You do not need to assign 0 to floats. 0.0 is a [zero value for `float64`](https://go.dev/ref/spec#The_zero_value).

File: pkg/scalers/splunk_o11y_scaler.go [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+		} else {
+			return nil, fmt.Errorf("no queryValue given")
+		}
+	}
+
+	// activationQueryValue
+	meta.activationQueryValue = 0
```
Comment (iwankgb) [inline pkg/scalers/splunk_o11y_scaler.go:93]: Ditto; 0.0 is zero value for `float64`.

File: pkg/scalers/splunk_o11y_scaler.go [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+	if val, ok := config.TriggerMetadata["queryAggregator"]; ok && val != "" {
+		queryAggregator := strings.ToLower(val)
+		switch queryAggregator {
+		case "max", "min", "avg":
+			meta.queryAggegrator = queryAggregator
+		default:
+			return nil, fmt.Errorf("queryAggregator value %s has to be one of 'max', 'min', or 'avg'.", queryAggregator)
+		}
+	} else {
+		meta.queryAggegrator = ""
```
Comment (iwankgb) [inline pkg/scalers/splunk_o11y_scaler.go:112]: Empty string is zero value for a `string` :)

File: pkg/scalers/splunk_o11y_scaler.go [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+	}
+
+	// accessToken
+	if accessToken, ok := config.AuthParams["accessToken"]; ok {
+		// cleanedAccessToken := strings.ReplaceAll(accessToken, "\n", "")
+		// cleanedAccessToken = strings.ReplaceAll(accessToken, "\r", "")
```
Comment (iwankgb) [inline pkg/scalers/splunk_o11y_scaler.go:118]: Is this cleaning still necessary?

File: pkg/scalers/splunk_o11y_scaler.go [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+
+	// Debug TODO check
+	meta.vType = v2.ValueMetricType
+
+	return &meta, nil
+}
+
+func newSplunkO11yConnection(ctx context.Context, meta *splunkO11yMetadata, config *scalersconfig.ScalerConfig) (*signalflow.Client, error) {
+	accessToken := meta.accessToken
+	realm := meta.realm
```
Comment (iwankgb) [inline pkg/scalers/splunk_o11y_scaler.go:140]: I don't think you need to assign these two variables; you can use properties of `meta` directly.

File: pkg/scalers/splunk_o11y_scaler.go [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+	}
+	logger.Info("Splunk Realm -> %s\n", meta.realm)
+
+	// Debug TODO check
+	meta.vType = v2.ValueMetricType
+
+	return &meta, nil
+}
+
+func newSplunkO11yConnection(ctx context.Context, meta *splunkO11yMetadata, config *scalersconfig.ScalerConfig) (*signalflow.Client, error) {
```
Comment (iwankgb) [inline pkg/scalers/splunk_o11y_scaler.go:138]: `config` seems to be unused in this function. Can you remove the argument, please?

File: pkg/scalers/splunk_o11y_scaler.go [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+
+func (s *splunkO11yScaler) getQueryResult(ctx context.Context) (float64, error) {
+	// var duration time.Duration = 1000000000 // one second in nano seconds
+	var duration time.Duration = 10000000000 // ten seconds in nano seconds
+
+	comp, err := s.apiClient.Execute(context.Background(), &signalflow.ExecuteRequest{
+		Program: s.metadata.query,
+	})
+	if err != nil {
+		return -1, fmt.Errorf("error: could not execute signalflow query: %w", err)
```
Comment (iwankgb) [inline pkg/scalers/splunk_o11y_scaler.go:173]: You should return zero value (0.0 in this case) when returning an error.

File: pkg/scalers/splunk_o11y_scaler.go [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+	}
+
+	go func() {
+		time.Sleep(duration)
+		if err := comp.Stop(context.Background()); err != nil {
+			s.logger.Info("Failed to stop computation")
+		}
+	}()
```
Comment (iwankgb) [inline pkg/scalers/splunk_o11y_scaler.go:181]: You should rather use [`WithTimeout()`](https://pkg.go.dev/context#WithTimeout):
```suggestion
	var duration time.Duration = 10000000000 // ten seconds in nano seconds

        xCtx, cancel := context.WithTimeout(ctx, duration)
        defer cancel()
	comp, err := s.apiClient.Execute(xCtx, &signalflow.ExecuteRequest{
		Program: s.metadata.query,
	})
	if err != nil {
		return 0.0, fmt.Errorf("error: could not execute signalflow query: %w", err)
	}
```

File: pkg/scalers/splunk_o11y_scaler.go [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+		return max, nil
+	case "min":
+		logMessage(s.logger, "Returning min value ", min)
+		return min, nil
+	case "avg":
+		avg := valueSum / float64(valueCount)
+		logMessage(s.logger, "Returning avg value ", avg)
+		return avg, nil
+	default:
+		return max, nil
```
Comment (iwankgb) [inline pkg/scalers/splunk_o11y_scaler.go:229]: Can these aggregations be performed as part of the Splunk query? I think it would be more reliable to use server side logic.

File: pkg/scalers/splunk_o11y_scaler.go [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+			if value > max {
+				max = value
+			}
+			if value < min {
+				min = value
+			}
+			valueSum += value
+			valueCount++
+		}
+	}
```
Comment (iwankgb) [inline pkg/scalers/splunk_o11y_scaler.go:211]: If you decide to use Splunk-side aggregation, then this loop would be significantly simpler and you should expect one and only one value.

File: pkg/scalers/splunk_o11y_scaler.go [hunk 16]
```diff
@@ -0,0 +1,264 @@
...
+	// s.logger.Info(fmt.Sprintf("splunk_o11y_scaler found authtrigger token : %s", s.metadata.accessToken))
+	num, err := s.getQueryResult(ctx)
+
+	if err != nil {
+		s.logger.Error(err, "error getting metrics from Splunk Observability Cloud.")
+		return []external_metrics.ExternalMetricValue{}, false, fmt.Errorf("error getting metrics from Splunk Observability Cloud: %w", err)
+	}
+	metric := GenerateMetricInMili(metricName, num)
+
+	logMessage(s.logger, "num", num)
```
Comment (iwankgb) [inline pkg/scalers/splunk_o11y_scaler.go:243]: `num` will never equal -1 here (-1 is returned from `s.getQueryResult()` only when error is returned and such a scenario id handled in line 237).

File: Makefile [hunk 17]
```diff
@@ -114,7 +114,7 @@ e2e-test-clean-crds: ## Delete all scaled objects and jobs across all namespaces
 
 .PHONY: e2e-test-clean
 e2e-test-clean: get-cluster-context ## Delete all namespaces labeled with type=e2e
-	kubectl delete ns -l type=e2e
+	microk8s kubectl delete ns -l type=e2e
```
[the remark this answers is not in the document]
Comment (sschimper-splunk) [reply Makefile:117]: Hi, thanks for pointing that out. I manually added "microk8s" to be able to build the project with my MicroK8s cluster that I have set up.

File: go.mod [hunk 18]
```diff
@@ -77,7 +77,7 @@ require (
 	github.com/segmentio/kafka-go/sasl/aws_msk_iam_v2 v0.1.0
 	github.com/spf13/cast v1.6.0
 	github.com/spf13/pflag v1.0.5
-	github.com/stretchr/testify v1.8.4
+	github.com/stretchr/testify v1.9.0
```
[the remark this answers is not in the document]
Comment (sschimper-splunk) [reply go.mod:80]: Understood, will do.

File: gomock_reflect_1996011792/prog.go [hunk 19]
```diff
@@ -0,0 +1,79 @@
+// Code generated by MockGen. DO NOT EDIT.
```
[the remark this answers is not in the document]
Comment (sschimper-splunk) [reply gomock_reflect_1996011792/prog.go:1]: I think it was created either automatically or already present. I don't have an idea of the purpose of this file right now, but I will look into it.

File: pkg/metricsservice/api/metrics.pb.go (hunk 1 above)
Comment (sschimper-splunk) [reply pkg/metricsservice/api/metrics.pb.go:19]: I am using version 25.2. Latest version is 28.0. I am not sure if 25.2 is outdated.

File: pkg/scalers/liiklus/LiiklusService.pb.go (hunk 2 above)
Comment (sschimper-splunk) [reply pkg/scalers/liiklus/LiiklusService.pb.go:4]: Yes, the file is unrelated to my changes.

File: pkg/scalers/liiklus/LiiklusService_grpc.pb.go (hunk 3 above)
Comment (sschimper-splunk) [reply pkg/scalers/liiklus/LiiklusService_grpc.pb.go:4]: yes.

File: pkg/scalers/splunk_o11y_scaler.go (hunk 4 above)
Comment (sschimper-splunk) [reply pkg/scalers/splunk_o11y_scaler.go:8]: Yes, I will do. Thank you for pointing this out to me.

File: pkg/scalers/splunk_o11y_scaler.go (hunk 9 above)
Comment (sschimper-splunk) [reply pkg/scalers/splunk_o11y_scaler.go:118]: No, I should have done a better job with the clean up :(

File: pkg/scalers/splunk_o11y_scaler.go (hunk 10 above)
Comment (sschimper-splunk) [reply pkg/scalers/splunk_o11y_scaler.go:140]: True. I created these variables for some print debugging.

File: pkg/scalers/splunk_o11y_scaler.go (hunk 11 above)
Comment (sschimper-splunk) [reply pkg/scalers/splunk_o11y_scaler.go:138]: Done.

File: pkg/scalers/splunk_o11y_scaler.go (hunk 13 above)
Comment (sschimper-splunk) [reply pkg/scalers/splunk_o11y_scaler.go:181]: Yes, that makes sense. I will update the function accordingly.

File: pkg/scalers/splunk_o11y_scaler.go (hunk 14 above)
Comment (sschimper-splunk) [reply pkg/scalers/splunk_o11y_scaler.go:229]: I will look into that.

**Comment by sschimper-splunk** [comment]
Hi @JorTurFer,
I will clarify this and get back to you.

**Comment by iwankgb** [comment]
> Is this still active?

I hope so :) Sebastian was off for a while as far as I understand.

**Comment by JorTurFer** [comment]
There is other PR merged supporting Splunk -> https://github.com/kedacore/keda/pull/5905
Is this a different feature or so?

**Comment by sschimper-splunk** [comment]
Hi @JorTurFer, this seems to be independent of our work on a Scaler for Splunk Observability.

**Comment by JorTurFer** [comment]
Could we extend the current Splunk scaler instead of creating a totally new one? (IDK if it's possible or not tbh)

**Comment by sschimper-splunk** [comment]
Yes, I'd even think its for the better.

**Comment by JorTurFer** [comment]
> Yes, I'd even think its for the better.

Thanks ❤️  by chance, 2 different contributions related with spluk have been done at the same time (this and the other already merged) xD.

We want to cut a release during the next week, if the PR is merged before Thursday, it'll be released next week (no rush at all, if it's not during this release, it'll be in the next one)

**Review by zroubalik — COMMENTED** [review]
Any updates on this please Shall we extend the current Splunk scaler?

**Comment by sschimper-splunk** [comment]
Hi @zroubalik, 
the work on a Splunk Keda scaler on the mentioned PR was regarding our Splunk Enterprise/Cloud solutions. I would like to work towards a Keda scaler for Splunk Observability Cloud.
If that is alright with you, I'd like to submit another PR in a "proper fashion", and therefore we can delete this PR.
Thank you, and apologies for all the confusion.
